intake_esm/cesm_aws.py

```diff
@@ -0,0 +1,122 @@
+
+            - Removing duplicates
+            - Adding extra metadata
+
+        Parameters
+        ----------
+        df_files : dict
+             Dictionary containing Pandas dataframes for different data sources
+
+
+        Returns
+        --------
+        df : pandas.DataFrame
+            Cleaned pandas dataframe containing all entries
+
+        """
+
+        df = pd.concat(list(df_files.values()), ignore_index=True, sort=False)
+        # Reorder columns
+        df = df[self.columns]
+
+        # Remove duplicates
+        df = df.drop_duplicates(subset=['resource', 'store_fullpath'], keep='last').reset_index(
+            drop=True
+        )
+        return df
+
+
+class CESMAWSSource(BaseSource):
+
+    name = 'cesm-aws'
+    partition_access = True
+
+    @staticmethod
+    def _validate_zarr_kwargs(kwargs):
+        _kwargs = {}
+        _kwargs['group'] = kwargs.get('group', None)
+        _kwargs['synchronizer'] = kwargs.get('synchronizer', None)
+        _kwargs['auto_chunk'] = kwargs.get('auto_chunk', True)
+        _kwargs['decode_cf'] = kwargs.get('decode_cf', True)
+        _kwargs['decode_times'] = kwargs.get('decode_times', True)
+        _kwargs['decode_coords'] = kwargs.get('decode_coords', True)
+        _kwargs['mask_and_scale'] = kwargs.get('mask_and_scale', True)
+        _kwargs['concat_characters'] = kwargs.get('concat_characters', True)
+        _kwargs['drop_variables'] = kwargs.get('drop_variables', None)
+        _kwargs['consolidated'] = kwargs.get('consolidated', True)
+        return _kwargs
+
+    def _open_dataset(self):
+        # fields which define a single dataset
+        dataset_fields = ['component', 'frequency']
+        kwargs = self._validate_kwargs(self.kwargs)
+        zarr_kwargs = CESMAWSSource._validate_zarr_kwargs(kwargs)
+
+        query_results = get_subset(self.collection_name, self.query)
+        grouped = query_results.groupby(dataset_fields)
+        all_dsets = {}
+        for dset_keys, dset_stores in tqdm(grouped, desc='dataset(s)'):
+            dset_id = '.'.join(dset_keys)
+            grouped_exp = dset_stores.groupby('experiment')
+            dsets = []
+            for exp_id, exp_stores in grouped_exp:
+                exp_dsets = []
+                for v_id, v_stores in tqdm(exp_stores.groupby('variable'), desc='variable(s)'):
+                    urlpath_ei_vi = v_stores['store_fullpath'].tolist()
+                    v_dsets = [
+                        aggregate.open_store(
+                            url,
+                            data_vars=[v_id],
+                            storage_options=self.storage_options,
+                            **zarr_kwargs,
+                        )
+                        for url in urlpath_ei_vi
+                    ]
+                    exp_dsets.extend(v_dsets)
+                exp_dset = aggregate.merge(exp_dsets)
+                dsets.append(exp_dset)
+
+            dset = aggregate.concat_time_levels(
```

Comment:
Are we concatenation along an ensemble dimension?

Reply:
Unlike the netcdf datasets, the zarr stores contains all the ensembles in one zarr store. Therefore, we don't need to the concatenation along the ensemble dimension.